api/v1beta1/openstackmachine_types.go

```diff
@@ -33,6 +33,59 @@ const (
 	IPClaimMachineFinalizer = "openstackmachine.infrastructure.cluster.x-k8s.io/ip-claim"
 )
 
+// SchedulerHintValueType is the type that represents allowed values for the Type field.
+// +kubebuilder:validation:Enum=Bool;String;Number
+type SchedulerHintValueType string
+
+// Constants representing the allowed types for SchedulerHintAdditionalValue.
+const (
+	SchedulerHintTypeBool   SchedulerHintValueType = "Bool"
+	SchedulerHintTypeString SchedulerHintValueType = "String"
+	SchedulerHintTypeNumber SchedulerHintValueType = "Number"
+)
+
+// SchedulerHintAdditionalValue represents the value of a scheduler hint property.
+// The value can be of various types: bool, string, or int.
+// The Type field indicates the type of the value being used.
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Bool' ? has(self.bool) : !has(self.bool)",message="bool is required when type is Bool, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'Number' ? has(self.number) : !has(self.number)",message="number is required when type is Number, and forbidden otherwise"
+// +kubebuilder:validation:XValidation:rule="has(self.type) && self.type == 'String' ? has(self.string) : !has(self.string)",message="string is required when type is String, and forbidden otherwise"
+// +union.
+type SchedulerHintAdditionalValue struct {
+	// Type represents the type of the value (Bool, String, or Number).
+	// +kubebuilder:validation:Required
+	// +unionDiscriminator
+	Type SchedulerHintValueType `json:"type"`
+
+	// Bool is the boolean value of the scheduler hint, used when Type is "Bool".
+	// This field is required if type is 'Bool', and must not be set otherwise.
+	// +unionMember,optional
+	Bool *bool `json:"bool,omitempty"`
+
+	// Number is the integer value of the scheduler hint, used when Type is "Number".
+	// This field is required if type is 'Number', and must not be set otherwise.
+	// +unionMember,optional
+	Number *int `json:"number,omitempty"`
+
+	// String is the string value of the scheduler hint, used when Type is "String".
+	// This field is required if type is 'String', and must not be set otherwise.
+	// +unionMember,optional
+	String *string `json:"string,omitempty"`
+}
+
+// SchedulerHintAdditionalProperty represents a single additional property for a scheduler hint.
+// It includes a Name to identify the property and a Value that can be of various types.
+type SchedulerHintAdditionalProperty struct {
+	// Name is the name of the scheduler hint property.
+	// It is a unique identifier for the property.
+	// +kubebuilder:validation:MinLength:=1
+	Name string `json:"name"`
+
+	// Value is the value of the scheduler hint property, which can be of various types
+	// (e.g., bool, string, int). The type is indicated by the Value.Type field.
```

Comment:
did we miss the validation?

```
// +kubebuilder:validation:Required
```

Reply:
It's implicit here, which is why the tests pass, but I agree that we should add it anyway. We've been trying to be explicit with these.